plugin/plugins/filesystem/nodes/ipfs.go

```diff
@@ -0,0 +1,219 @@
+	metaClone.Filter(req)
+
+	return id.Qid, req, metaClone, nil
+}
+
+func (id *IPFS) Walk(names []string) ([]p9.QID, p9.File, error) {
+	id.Logger.Debugf("ID Walk names %v", names)
+	id.Logger.Debugf("ID Walk myself: %s:%v", id.Path, id.Qid)
+
+	if shouldClone(names) {
+		id.Logger.Debugf("ID Walk cloned")
+		return []p9.QID{id.Qid}, id, nil
+	}
+
+	qids := make([]p9.QID, 0, len(names))
+
+	//TODO: [spec check] make sure we're not messing things up by doing this instead of mutating
+	// ^ Does internal library expect fid to mutate on success or does newfid clobber some external state anyway
+	walkedNode := &IPFS{} // operate on a copy
+	*walkedNode = *id
+
+	var err error
+	for _, name := range names {
+		//TODO: timerctx; don't want to hang forever
+		if walkedNode.Path, err = id.core.ResolvePath(id.Ctx, corepath.Join(walkedNode.Path, name)); err != nil {
+			return qids, nil, err
+		}
+
+		ipldNode, err := id.core.Dag().Get(id.Ctx, walkedNode.Path.Cid())
+		if err != nil {
+			return qids, nil, err
+		}
+
+		//TODO: this is too opague; we want core path => qid, dirent isn't necessary
+		dirEnt := &p9.Dirent{}
+		if err = ipldStat(dirEnt, ipldNode); err != nil {
+			return qids, nil, err
+		}
+
+		walkedNode.Qid = dirEnt.QID
+		//
+		qids = append(qids, walkedNode.Qid)
+	}
+
+	id.Logger.Debugf("ID Walk ret %v, %v", qids, walkedNode)
+	return qids, walkedNode, err
+}
+
+func (id *IPFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	id.Logger.Debugf("ID Open")
+	if id.meta.Mode.IsDir() {
+		c, err := id.core.Unixfs().Ls(id.Ctx, id.Path)
+		if err != nil {
+			return id.Qid, 0, err
+		}
+
+		id.directory = &directoryStream{
+			entryChan: c,
+		}
+		return id.Qid, 0, nil
+	}
+
+	// else treat as a file
+	apiNode, err := id.core.Unixfs().Get(id.Ctx, id.Path)
+	if err != nil {
+		return id.Qid, 0, err
+	}
+
+	var ok bool
+	if id.file, ok = apiNode.(files.File); !ok {
+		return id.Qid, 0, fmt.Errorf("%q does not appear to be a file: %T", id.Path.String(), apiNode)
+	}
+
+	return id.Qid, ipfsBlockSize, nil
+}
+
+func (id *IPFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	id.Logger.Debugf("ID Readdir %d %d", offset, count)
+
+	if id.directory == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", id.Path.String())
+	}
+
+	if offset < 0 {
+		return nil, fmt.Errorf("offset %d can't be negative", offset)
+	}
+
+	if id.directory.eos {
+		if offset == id.directory.cursor-1 {
+			return nil, nil // valid end of stream request
+		}
+	}
+
+	if id.directory.eos && offset == id.directory.cursor-1 {
+		return nil, nil // valid end of stream request
+	}
+
+	if offset < id.directory.cursor {
+		return nil, fmt.Errorf("read offset %d is behind current entry %d, seeking backwards in directory streams is not supported", offset, id.directory.cursor)
+	}
+
+	ents := make([]p9.Dirent, 0)
+out:
+	for len(ents) < int(count) {
+		select {
+		case entry, open := <-id.directory.entryChan:
+			if !open {
+				id.directory.eos = true
```

Comment:
This works but I'm not happy about it.
We need some way to tell we've hit the end of the stream, so when we get a `Readdir` request containing an `offset` of  `entryCount+1`, we can return nothing+success (nil, nil), signifying the end of the stream.

Consider a directory containing 7 entries.
Linux's `ls` will request `Readdir(0, $largeNumber)`, to which we return 7 entries.
The last entry will contain 0-index offset of 6. Linux will request from us now `Readdir(6, $largeNumber)`, expecting (nil, nil)

If we have a way of retrieving the entry count on `Open`, we could just compare the offset against it to see if we're at/past the end. Previously we used the IPFS object's `NumLinks` attribute, but this is not valid for sharded directories.

Reply:
Do we _have_ to remember the EOS? We should be able to just read from the channel. If the channel is closed, we're at the end of the stream.

We might also need to return an error if `offset > cursor` but I'm not sure what the spec says about seeking past the end of the directory.

Reply:
Currently, this only serves to handle the requests where the offset points to the last entry.

Since the cursor will be at `$entryCount+1` on a full read (https://github.com/ipfs/go-ipfs/pull/6612/files#r325418329),
and the request offset will be equal to `$entryCount`, we need to make an exception, allowing the offset to be behind the cursor, (only) if we've successfully hit the end already.
In all other scenarios, if the offset is behind the cursor, we treat it as invalid.

Same as before, there may be a better way of handling this, but I'm unsure of it at the moment.

Reply:
That multiple kinds of messed up. This should be better.
https://github.com/ipfs/go-ipfs/pull/6612/commits/7bb94a4da28eb8a6a4b322a35128264d07891711

Reply:
@Stebalien 
Requesting another pass on this area specifically.

Reply:
Also highlighting the upper bound check mentioned by @hugelgupf
https://github.com/hugelgupf/p9/pull/12#discussion_r337324164

We may want to incorporate that for our channel reader here too.